File: backend/src/domain/events.ts
import {
  Actor,
  assertSafeData,
  normalizeUtc,
  utcNow,
  validateJsonObject,
  validateSafeValue,
} from "./common.js";
import { newObjectId } from "./common.js";

/** 调用/安全事件的正式类型集合；这类事件必须携带 attempt、actor 和脱敏原因。 */
const CONTEXT_EVENT_TYPES = new Set([
  "InvocationStarted",
  "InvocationCompleted",
  "InvocationFinished",
  "InvocationFailed",
  "InvocationRejected",
  "InvocationSecurity",
  "InvocationPolicy",
  "ModelCallStarted",
  "ModelCallCompleted",
  "ModelCallFinished",
  "ModelCallFailed",
  "ModelCallRejected",
  "ToolCallStarted",
  "ToolCallCompleted",
  "ToolCallFinished",
  "ToolCallFailed",
  "ToolCallRejected",
]);
/** 待提交的追加领域事实。 */
export type DomainEventDraft = {
  eventType: string;
  aggregateType: string;
  aggregateId: string;
  aggregateVersion?: number;
  payload: Record<string, unknown>;
  inputSummary: string | Record<string, unknown>;
  outputSummary: string | Record<string, unknown>;
  result: string;
  failure: string | null;
  retryCount: number;
  durationMs: number;
  actor: Actor;
  traceId: string;
  occurredAt: string;
  attemptId: string | null;
  rejectionReason: string | null;
  redactionReason: string | null;
  eventCategory: "ordinary" | "call" | "security";
};
/** 已定位到聚合版本的不可变领域事实。 */
export type DomainEvent = DomainEventDraft & {
  eventId: string;
  aggregateVersion: number;
  globalSequence: number;
};
/** 描述一批追加事件的聚合版本结果。 */
export type AppendResult = {
  aggregateType: string;
  aggregateId: string;
  expectedVersion: number;
  aggregateVersion: number;
  events: DomainEvent[];
};

/** 校验领域事件，防止敏感内容和上下文事件缺少追溯字段。 */
export function parseEventDraft(
  input: Partial<DomainEventDraft> & Record<string, unknown>,
): DomainEventDraft {
  const eventType = validateSafeValue(
    String(input.eventType ?? ""),
    "eventType",
  );
  const category = (input.eventCategory ??
    "ordinary") as DomainEventDraft["eventCategory"];
  if (!["ordinary", "call", "security"].includes(category))
    throw new Error("invalid eventCategory");
  const attemptId =
    input.attemptId == null
      ? null
      : validateSafeValue(String(input.attemptId), "attemptId");
  const actorInput = input.actor;
  if (!actorInput || typeof actorInput !== "object")
    throw new Error("actor is required");
  const actor = {
    type: validateSafeValue(String((actorInput as Actor).type), "actor.type"),
    id: validateSafeValue(String((actorInput as Actor).id), "actor.id"),
  };
  const rejectionReason =
    input.rejectionReason == null ? null : String(input.rejectionReason);
  const redactionReason =
    input.redactionReason == null ? null : String(input.redactionReason);
  assertSafeData(failureSummary(input.failure));
  assertSafeData(rejectionReason ?? "");
  assertSafeData(redactionReason ?? "");
  if (
    (category !== "ordinary" || CONTEXT_EVENT_TYPES.has(eventType)) &&
    (!attemptId || !rejectionReason?.trim() || !redactionReason?.trim())
  )
    throw new Error(
      "context event requires: attemptId, actor, rejectionReason, redactionReason",
    );
  const payload = validateJsonObject(input.payload ?? {});
  assertSafeData(payload);
  assertSafeData(input.inputSummary ?? "");
  assertSafeData(input.outputSummary ?? "");
  return {
    eventType,
    aggregateType: validateSafeValue(
      String(input.aggregateType ?? ""),
      "aggregateType",
    ),
    aggregateId: validateSafeValue(
      String(input.aggregateId ?? ""),
      "aggregateId",
    ),
    aggregateVersion: nonNegative(
      input.aggregateVersion ?? input.expectedVersion ?? 0,
      "aggregateVersion",
    ),
    payload,
    inputSummary: (input.inputSummary ??
      "") as DomainEventDraft["inputSummary"],
    outputSummary: (input.outputSummary ??
      "") as DomainEventDraft["outputSummary"],
    result: validateSafeValue(String(input.result ?? "success"), "result"),
    failure: failureSummary(input.failure),
    retryCount: nonNegative(input.retryCount ?? 0, "retryCount"),
    durationMs: nonNegative(input.durationMs ?? 0, "durationMs"),
    actor,
    traceId: validateSafeValue(
      String(input.traceId ?? "trace_unknown"),
      "traceId",
    ),
    occurredAt: input.occurredAt
      ? normalizeUtc(String(input.occurredAt))
      : utcNow(),
    attemptId,
    rejectionReason,
    redactionReason,
    eventCategory: category,
  };
}

/** 确认事件的版本数量与聚合最终版本一致。 */
export function validateAppendResult(result: AppendResult): AppendResult {
  if (result.aggregateVersion !== result.expectedVersion + result.events.length)
    throw new Error(
      "aggregateVersion must equal expectedVersion plus event count",
    );
  return result;
}
/** 把待提交事件物化为带全局序号的不可变事件。 */
export function materializeEvent(
  draft: DomainEventDraft,
  aggregateVersion: number,
  globalSequence: number,
): DomainEvent {
  return {
    ...draft,
    aggregateVersion,
    globalSequence,
    eventId: newObjectId("event"),
  };
}
function nonNegative(value: unknown, name: string): number {
  if (!Number.isInteger(value) || Number(value) < 0)
    throw new Error(`${name} must be a non-negative integer`);
  return Number(value);
}

/** 规范化可选失败摘要，避免 undefined 或非字符串值进入事件事实。 */
function failureSummary(value: unknown): string | null {
  if (value == null) {
    return null;
  }

  if (typeof value !== "string") {
    throw new Error("failure must be a string or null");
  }

  return value;
}

File: backend/app/domain/events.py
"""Task 2 领域事件只追加合同和 EventStore 协议。

本模块不实现数据库写入；它只保证事件载荷可序列化、事件事实不可变，并为
后续仓储提供带 expected_version 的异步接口边界。
"""

from __future__ import annotations

from typing import Any, Literal, Optional, Protocol, Sequence, Union, runtime_checkable

from pydantic import AliasChoices, ConfigDict, Field, field_validator, model_validator

from .common import Actor, utc_now
from .entities import AwareDateTime, ActorValue, DomainModel, JsonObject, NonEmptyStr, NonNegativeInt, SafeSummary, Version


SummaryValue = Union[JsonObject, SafeSummary]
EventCategory = Literal["ordinary", "call", "security"]
_REGISTERED_CONTEXT_EVENT_TYPES = frozenset(
    {
        "InvocationStarted",
        "InvocationCompleted",
        "InvocationFinished",
        "InvocationFailed",
        "InvocationRejected",
        "InvocationSecurity",
        "InvocationPolicy",
        "InvocationSecurityBlocked",
        "InvocationPolicyBlocked",
        "ProviderInvocationStarted",
        "ProviderInvocationCompleted",
        "ProviderInvocationFinished",
        "ProviderInvocationFailed",
        "ProviderInvocationRejected",
        "ProviderInvocationSecurity",
        "ProviderInvocationPolicy",
        "ProviderInvocationSecurityBlocked",
        "ProviderInvocationPolicyBlocked",
        "ModelCallStarted",
        "ModelCallCompleted",
        "ModelCallFinished",
        "ModelCallFailed",
        "ModelCallRejected",
        "ModelCallSecurity",
        "ModelCallPolicy",
        "ModelCallSecurityBlocked",
        "ModelCallPolicyBlocked",
        "ToolCallStarted",
        "ToolCallCompleted",
        "ToolCallFinished",
        "ToolCallFailed",
        "ToolCallRejected",
        "ToolCallSecurity",
        "ToolCallPolicy",
        "ToolCallSecurityBlocked",
        "ToolCallPolicyBlocked",
    }
)
_REGISTERED_CONTEXT_EVENT_TYPES_CASEFOLDED = frozenset(
    event_type.casefold() for event_type in _REGISTERED_CONTEXT_EVENT_TYPES
)


def _safe_event_value(value: Any) -> Any:
    """拒绝事件摘要和小型 payload 中的凭据、提示词和令牌字段。"""
    from .entities import _freeze_json_value, _safe_data

    return _freeze_json_value(_safe_data(value))


class DomainEventDraft(DomainModel):
    """描述待提交的领域事实，不包含大型证据正文或敏感信息。"""

    model_config = ConfigDict(
        extra="forbid",
        populate_by_name=True,
        strict=True,
        frozen=True,
        validate_assignment=False,
        validate_default=True,
    )

    event_type: NonEmptyStr = Field(alias="eventType")
    aggregate_type: NonEmptyStr = Field(alias="aggregateType")
    aggregate_id: NonEmptyStr = Field(alias="aggregateId")
    aggregate_version: NonNegativeInt = Field(
        default=0,
        validation_alias=AliasChoices(
            "aggregate_version", "aggregateVersion", "expected_version", "expectedVersion"
        ),
        serialization_alias="aggregateVersion",
    )
    payload: JsonObject = Field(default_factory=dict)
    input_summary: SummaryValue = Field(default="", alias="inputSummary")
    output_summary: SummaryValue = Field(default="", alias="outputSummary")
    result: NonEmptyStr = "success"
    failure: Optional[NonEmptyStr] = None
    retry_count: NonNegativeInt = 0
    duration_ms: NonNegativeInt = 0
    actor: ActorValue = Field(default_factory=lambda: Actor(type="system", id="domain"))
    trace_id: NonEmptyStr = Field(default="trace_unknown", alias="traceId")
    occurred_at: AwareDateTime = Field(default_factory=utc_now, alias="occurredAt")
    attempt_id: Optional[NonEmptyStr] = Field(default=None, alias="attemptId")
    rejection_reason: Optional[SafeSummary] = Field(default=None, alias="rejectionReason")
    redaction_reason: Optional[SafeSummary] = Field(default=None, alias="redactionReason")
    event_category: EventCategory = Field(default="ordinary", alias="eventCategory")

    _check_event_values = field_validator(
        "input_summary",
        "output_summary",
        "failure",
        "payload",
        "rejection_reason",
        "redaction_reason",
        mode="after",
    )(_safe_event_value)

    @model_validator(mode="after")
    def validate_context_event_contract(self) -> "DomainEventDraft":
        """安全/调用事件必须带执行尝试、显式操作者和脱敏原因。"""
        event_type = self.event_type.casefold()
        if (
            event_type not in _REGISTERED_CONTEXT_EVENT_TYPES_CASEFOLDED
            and self.event_category == "ordinary"
        ):
            return self

        missing: list[str] = []
        if self.attempt_id is None:
            missing.append("attempt_id")
        if "actor" not in self.model_fields_set:
            missing.append("actor")
        if self.rejection_reason is None or not self.rejection_reason.strip():
            missing.append("rejection_reason")
        if self.redaction_reason is None or not self.redaction_reason.strip():
            missing.append("redaction_reason")
        if missing:
            raise ValueError(
                "context event requires: " + ", ".join(missing)
            )
        return self


class DomainEvent(DomainEventDraft):
    """保存已经定位到聚合和版本的不可变领域事实。"""

    event_id: NonEmptyStr = Field(alias="eventId")
    aggregate_version: Version = Field(
        validation_alias=AliasChoices(
            "aggregate_version", "aggregateVersion", "expected_version", "expectedVersion"
        ),
        serialization_alias="aggregateVersion",
    )
    global_sequence: NonNegativeInt = Field(default=0, alias="globalSequence")


class AppendResult(DomainModel):
    """描述一次事件追加的聚合版本结果，供后续事务仓储返回。"""

    aggregate_type: NonEmptyStr
    aggregate_id: NonEmptyStr
    expected_version: NonNegativeInt
    aggregate_version: NonNegativeInt
    events: tuple[DomainEvent, ...]

    @model_validator(mode="after")
    def validate_version_sequence(self) -> "AppendResult":
        """确保事件数量和首尾版本不会在合同层自相矛盾。"""
        if self.aggregate_version != self.expected_version + len(self.events):
            raise ValueError("aggregate_version must equal expected_version plus event count")
        return self


@runtime_checkable
class EventStore(Protocol):
    """定义后续 SQLAlchemy 事件仓储需要实现的追加和游标读取接口。"""

    async def append(
        self,
        aggregate_type: str,
        aggregate_id: str,
        expected_version: int,
        events: Sequence[DomainEventDraft],
    ) -> AppendResult:
        """原子追加一组事件并校验聚合 expected_version。"""
        ...

    async def list_after(
        self,
        event_id: Optional[str],
        project_id: Optional[str] = None,
    ) -> list[DomainEvent]:
        """按稳定 event_id 游标读取事件，不把游标解释为 offset。"""
        ...
